### PR TITLE
Compare exact member of ISR with RS when deciding whether the partition is URP

### PR DESCRIPTION
**Pull request by suyashtava**
This PR resolves [#1905 ](https://github.com/linkedin/cruise-control/issues/1905)

Today, the way we determine a partition is an URP (Under Replicated Partition) based on the size of ISR and RS. E.g. if ISR = [A, B], and RS = [A, B, C], then the partition is treated as URP by CC.

However, in certain cases (usually because kafka is in a bad/inconsistent state), the ISR can be diverged from RS. E.g. ISR = [A, B, B], while RS = [D, E, F]. In this case, we still want CC to report this replica as URP, since the partition is not in a good state.

The proposal is, instead of comparing the size of ISR and RS, we should comparing the members in ISR and RS. If ISR != RS, CC should report it as URP.

### CODE REVIEW

**Comment by suyashtava** [comment]
CC: @CCisGG

**Comment by CCisGG** [comment]
@suyashtava that is a flaky test. I can re-trigger it.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ClusterPartitionState.java [hunk 1]
```diff
@@ -112,7 +112,7 @@ protected void populateKafkaPartitionState(Set<PartitionInfo> underReplicatedPar
         }
         for (PartitionInfo partitionInfo : _kafkaCluster.partitionsForTopic(topic)) {
           int numInsyncReplicas = partitionInfo.inSyncReplicas().length;
-          boolean isURP = numInsyncReplicas != partitionInfo.replicas().length;
+          boolean isURP = !(partitionInfo.inSyncReplicas() == partitionInfo.replicas());
```
Comment (CCisGG) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ClusterPartitionState.java:115]: I don't think this is safe enough. E.g. if the ISR set is same as replica set, but they are not the same order, then isURP will be true. But I think in that case, isURP should be false.

I feel we should compare the two Node[] in a sorted manner

**Comment by suyashtava** [comment]
Not very much aligned because as far as I have seen. Replicas always come
in this order. {Leader/isr1, Isr2/follower1, isr3/follower2}

While ISR comes in {isr1,isr2,isr3} considering all 3 are isr.

So they will always maintain the order.

If there’s a discrepancy in order then this is not a normal behaviour, and
should be called underrep.

@CCisGG What are your thoughts?

But if we are ignoring this abnormal behaviour than we can go for sorting.

**Comment by CCisGG** [comment]
@suyashtava I just checked our cluster, and the order in ISR set is not what you listed. The leader is not always the first one. We should compare the two only after sorting.

**Comment by CCisGG** [comment]
Besides, could you make the PR title more explicit and concrete? The PR title will be automatically convert to the commit message by default once merged. We should make sure the commit message is clear and informative.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ClusterPartitionState.java [hunk 2]
```diff
@@ -112,7 +113,12 @@ protected void populateKafkaPartitionState(Set<PartitionInfo> underReplicatedPar
         }
         for (PartitionInfo partitionInfo : _kafkaCluster.partitionsForTopic(topic)) {
           int numInsyncReplicas = partitionInfo.inSyncReplicas().length;
-          boolean isURP = !(partitionInfo.inSyncReplicas() == partitionInfo.replicas());
+          Node[] inSyncReplicas = Arrays.copyOf(partitionInfo.inSyncReplicas(), numInsyncReplicas);
+          Arrays.sort(inSyncReplicas);
+          int numReplicas = partitionInfo.replicas().length;
+          Node[] replicas = Arrays.copyOf(partitionInfo.replicas(), numReplicas);
+          Arrays.sort(replicas);
+          boolean isURP = !(replicas == inSyncReplicas);
```
Comment (CCisGG) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ClusterPartitionState.java:121]: "==" looks not right here, as it compares the reference, not the array content. I think Arrays.equal should be used here. That said, I think we can extract a method to compare between ISR and replica set, and unit test it. That way we can feel safer before ship it.

Comment (suyashtava) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ClusterPartitionState.java:121]: Yup, Array Equals makes more sense And Unit test aswell.

**Comment by suyashtava** [comment]
Hi @CCisGG , I checked the Test file: There no any such Unit test here for any helper function([example](https://github.com/linkedin/cruise-control/blob/b68431fbe303558ca3868ba802ce2663690344c0/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ClusterPartitionState.java#L145)) in [Response.test](https://github.com/linkedin/cruise-control/blob/migrate_to_kafka_2_5/cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/servlet/response/ResponseTest.java)

I would suggest let's go ahead with this, as it is not such a big change and I can create a new PR/Issue, where we can create all the missing unit tests, in this or any file. That would go a long way.

Your thoughts?

**Comment by suyashtava** [comment]
Further was thinking will it make sense to have a response until file and we can move all our helpers functions there.

And then have unit test for all helpers functions in that file. That will be more systematic

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ClusterPartitionState.java [hunk 3]
```diff
@@ -83,6 +86,23 @@ public Map<String, Object> getJsonStructure() {
...
+   * @return the effective boolean under replicated status of a partition.
+   */
+  protected boolean isURP(PartitionInfo partitionInfo) {
+    int numInsyncReplicas = partitionInfo.inSyncReplicas().length;
+    Node[] inSyncReplicas = Arrays.copyOf(partitionInfo.inSyncReplicas(), numInsyncReplicas);
+    Arrays.sort(inSyncReplicas);
+    int numReplicas = partitionInfo.replicas().length;
+    Node[] replicas = Arrays.copyOf(partitionInfo.replicas(), numReplicas);
+    Arrays.sort(replicas);
+    boolean isURP = !(Arrays.equals(replicas, inSyncReplicas));
```
Comment (CCisGG) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ClusterPartitionState.java:103]: Redundant brackets here. "!Arrays.equals(replicas, inSyncReplicas);" should be good enough.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ClusterPartitionState.java [hunk 4]
```diff
@@ -112,7 +132,7 @@ protected void populateKafkaPartitionState(Set<PartitionInfo> underReplicatedPar
         }
         for (PartitionInfo partitionInfo : _kafkaCluster.partitionsForTopic(topic)) {
           int numInsyncReplicas = partitionInfo.inSyncReplicas().length;
```
Comment (CCisGG) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ClusterPartitionState.java:134]: Do we still need this var?

**Review by CCisGG — CHANGES_REQUESTED** [review]
Changes look good to me. I'm sort of worrying about the perf impact when there are many partitions. We can merge it and we will see how does it work on our large clusters.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ClusterPartitionState.java (hunk 4 above)
Comment (suyashtava) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ClusterPartitionState.java:134]: Yes, we still need it to check if th3e current Partition is underMinIsrPartitions

Comment (suyashtava) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/ClusterPartitionState.java:134]: Resolving this as of Now.

**Comment by suyashtava** [comment]
Updated the redundant brackets.

Btw do you want me to work on creating a utils file and move all  helper functions there and unit tests for each of them?

**Comment by CCisGG** [comment]
@suyashtava That would be great and helpful. Would you like to do it in this same PR or you want a new one?

**Comment by suyashtava** [comment]
I would prefer we should do it seprately as a new WI.

That way we can track things down better which helpers functions need to be
moved and we will easier to track down mistake as it won’t create functions
in case of things break.


Should I create a new issue for the same, or directly create a placeholder
PR and start working there.

**Comment by CCisGG** [comment]
@suyashtava a new issue would always be great.

**Review by CCisGG — APPROVED** [review]

**Comment by suyashtava** [comment]
Sure will create new issue and start working on it.
 Will do it in Day time of IST in morning, tomorrow.

We are good to close this one.

**Comment by CCisGG** [comment]
Hi @suyashtava , I have to revert the PR as it fails the integration test. 
https://app.circleci.com/pipelines/github/linkedin/cruise-control/1999/workflows/27fdca20-5672-4237-9c0c-a4a4f1c6cdcf/jobs/5725

Root cause is Node class cannot be cast to Comparable class. 

This is why we should add unit test for any changes. I'd say this time let's add the test before merge the PR.

**Pull request merged**